### PR TITLE
NXP-21561: MongoDB sequencer

### PR DESCRIPTION
**Pull request by kevinleturc**

### CODE REVIEW

**Comment by funshodavid** [comment]
:+1:

File: nuxeo-mongodb-core/src/main/java/org/nuxeo/mongodb/core/MongoDBComponent.java [hunk 1]
```diff
@@ -0,0 +1,93 @@
...
+import org.nuxeo.runtime.model.ComponentContext;
+import org.nuxeo.runtime.model.ComponentInstance;
+import org.nuxeo.runtime.model.DefaultComponent;
+
+import com.mongodb.MongoClient;
+import com.mongodb.client.MongoDatabase;
+
+/**
+ * Component used to get a database connection to MongoDB. Don't expose {@link MongoClient} directly, because it's this
+ * component which is responsible to create and close it.
```
Comment (efge) [inline nuxeo-mongodb-core/src/main/java/org/nuxeo/mongodb/core/MongoDBComponent.java:35]: responsible for creating and closing it

File: nuxeo-mongodb-seqgen/src/main/resources/OSGI-INF/mongodb-seqgen-contrib.xml [hunk 2]
```diff
@@ -0,0 +1,19 @@
+<?xml version="1.0"?>
+<component name="org.nuxeo.mongodb.seqgen.contrib">
+
+  <documentation>
+    MongoDB Audit Sesuence Generator uses MongoDBComponent. MongoDBComponent should be configured in order to have
```
Comment (efge) [inline nuxeo-mongodb-seqgen/src/main/resources/OSGI-INF/mongodb-seqgen-contrib.xml:5]: Sesuence -> Sequence

File: nuxeo-mongodb-seqgen/src/main/java/org/nuxeo/mongodb/seqgen/MongoDBUIDSequencer.java [hunk 3]
```diff
@@ -0,0 +1,101 @@
...
+import org.nuxeo.runtime.services.config.ConfigurationService;
+
+import com.mongodb.client.MongoCollection;
+import com.mongodb.client.MongoDatabase;
+import com.mongodb.client.model.Filters;
+import com.mongodb.client.model.FindOneAndUpdateOptions;
+import com.mongodb.client.model.Updates;
+
+/**
+ * MongoDB implementation of {@link UIDSequencer}.<br />
```
Comment (efge) [inline nuxeo-mongodb-seqgen/src/main/java/org/nuxeo/mongodb/seqgen/MongoDBUIDSequencer.java:37]: Use `<p>` not `<br/>`

File: nuxeo-mongodb-seqgen/src/main/java/org/nuxeo/mongodb/seqgen/MongoDBUIDSequencer.java [hunk 4]
```diff
@@ -0,0 +1,101 @@
...
+    @Override
+    public int getNext(String key) {
+        return (int) getNextLong(key);
+    }
+
+    @Override
+    public long getNextLong(String key) {
+        FindOneAndUpdateOptions options = new FindOneAndUpdateOptions().upsert(true);
+        Bson filter = Filters.eq(MongoDBSerializationHelper.MONGODB_ID, key);
+        Bson update = Updates.combine(Updates.inc("sequence", ONE),
```
Comment (efge) [inline nuxeo-mongodb-seqgen/src/main/java/org/nuxeo/mongodb/seqgen/MongoDBUIDSequencer.java:83]: `"sequence"` -> use a constant

File: nuxeo-mongodb-seqgen/src/main/java/org/nuxeo/mongodb/seqgen/MongoDBUIDSequencer.java [hunk 5]
```diff
@@ -0,0 +1,101 @@
...
+                Updates.setOnInsert(MongoDBSerializationHelper.MONGODB_ID, key));
+        Document sequence = getCollection().findOneAndUpdate(filter, update, options);
+        // If sequence is null, then it means we just create it - convert null to 1
+        if (sequence == null) {
+            return 1L;
+        }
+        // As we retrieve the document before the update we need to add 1 manually
+        return (long) MongoDBSerializationHelper.bsonToFieldMap(sequence).get("sequence") + 1;
```
Comment (efge) [inline nuxeo-mongodb-seqgen/src/main/java/org/nuxeo/mongodb/seqgen/MongoDBUIDSequencer.java:91]: use constant

**Comment by kevinleturc** [comment]
👍

File: nuxeo-mongodb-core/src/main/resources/OSGI-INF/mongodb-core-contrib.xml [hunk 6]
```diff
@@ -0,0 +1,24 @@
...
+    <documentation>
+      Extension point used to configure access to a MongoDB server or cluster.
+
+      This setup a MongoClient, server are mandatory.
+
+      <code>
+        <connection>
+          <server>mongodb://bob:pass@localhost:27017,otherhost:27018/mydb?replicaSet=test;connectTimeoutMS=300000</server>
+          <dbname>mydb</dbname>
+        </connection>
```
Comment (efge) [inline nuxeo-mongodb-core/src/main/resources/OSGI-INF/mongodb-core-contrib.xml:16]: As it is there can be only one connection provided by the `MongoDBComponent` in the whole Nuxeo instance. I think this is too restrictive and that you should configure an id for each, `<connection id="default">` for instance, and have the users of the `MongoDBComponent` specify which one they want.

**Pull request merged**